cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala

```diff
@@ -0,0 +1,206 @@
+package com.twitter.cassovary.graph
+
+import com.twitter.cassovary.graph.node.DynamicNode
+import it.unimi.dsi.fastutil.ints.IntArrayList
+import scala.collection.mutable.ArrayBuffer
+import StoredGraphDir._
+import com.twitter.cassovary.util.FastUtilUtils
+
+/**
+ * A dynamic directed graph, implemented using an ArrayBuffer of IntArrayList
+ * (from the fastutil library).  If n nodes are used, O(n) objects are created,
+ * independent of the number of edges.
+ *
+ * For efficiency, it's recommended that ids be sequentially numbered, as
+ * an internal array is stored which is longer than the maximum id seen.
+ *
+ * This class is not threadsafe.  If multiple thread access it simultaneously
+ * and at least one mutates it, synchronization should be used.
+ */
+class ArrayBasedDynamicDirectedGraph(override val storedGraphDir: StoredGraphDir)
+    extends DynamicDirectedGraph {
+  // outboundLists(id) contains the outbound neighbors of the given id,
+  // or null if the id is not in this graph.
+  // If we aren't storing outbound neighbors, outboundLists will always remain size 0.
+  private val outboundLists = new ArrayBuffer[IntArrayList]
+  // (See above note on outboundLists)
+  private val inboundLists = new ArrayBuffer[IntArrayList]
+
+  private var nodeCount_ = 0
+
+  def this(dataIterator: Iterator[NodeIdEdgesMaxId],
+            storedGraphDir: StoredGraphDir) {
+    this(storedGraphDir)
+    for (nodeData <- dataIterator) {
+      val id = nodeData.id
+      getOrCreateNode(id)
+      nodeData.edges map getOrCreateNode
+      storedGraphDir match {
+        case OnlyOut => outboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case OnlyIn => inboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case BothInOut => nodeData.edges map { addEdgeAllowingDuplicates(id, _) } // Duplicates shouldn't exist, but allow them for efficiency.
+      }
+    }
+  }
+
+  def this(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+           storedGraphDir: StoredGraphDir) {
+    this((iteratorSeq map { _()} ).iterator.flatten, storedGraphDir)
+  }
+
+  /* Returns an option which is non-empty if outbound list for id  is non-null. */
+  private def outboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= outboundLists.size)
+      None
+  else
+      Option(outboundLists(id)) // If list is null, convert to None
+
+  private def inboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= inboundLists.size)
+      None
+    else
+      Option(inboundLists(id)) // If list is null, convert to None
+
+  /** Wraps outboundList and inboundList in a node structure.  This is an inner class
+    * because mutations will affect other nodes of the graph if both inbound
+    * and outbound neighbor lists are stored.
+    *
+    * For efficiency, we don't store Nodes, but create them on the fly as needed.
+    */
+  class IntListNode(override val id: Int,
```

Comment:
If we want to avoid the storage of one of the two option classes, I think we would need to not access it inside the relevant closure, which would make the code awkward (I'm not sure how to do that without essentially writing two different Node classes). More precicely, in your code
`new DynamicNode {
....
}`
I assume the "...." contains references to both outboundList and inboundList, inless there are separate calls to `new Dynamic Node {...}` for each case.  Since Nodes aren't being stored anyway, I'm inclinded to leave it as is unless you think it's important.